onnxruntime/core/providers/cpu/quantization/quantize_linear.cc

```diff
@@ -439,6 +529,72 @@ void ComputeLoop(OpKernelContext* ctx, const InT* input, const InT* scale, const
   }
 }
 
+// Quantizes float32 to INT4 (in-place) using MLAS kernel.
+#define DEFINE_COMPUTE_LOOP_FP32_TO_INT4(INT4_TYPE, QUANT_FUNC)                                               \
+  template <>                                                                                                 \
+  void ComputeLoop(OpKernelContext* ctx, const float* input, const float* scale, const INT4_TYPE* zero_point, \
+                   INT4_TYPE* output, int64_t N, int64_t broadcast_dim, int64_t block_size, bool saturate) {  \
+    ORT_UNUSED_PARAMETER(saturate);                                                                           \
+    size_t output_index = 0;                                                                                  \
+    for (size_t n = 0; n < static_cast<size_t>(N); n++) {                                                     \
+      for (size_t bd = 0; bd < static_cast<size_t>(broadcast_dim); bd++) {                                    \
+        size_t bd_i = bd >> 1;  /*bd / 2*/                                                                    \
+        size_t bd_j = bd & 0x1; /*bd % 2*/                                                                    \
+        INT4_TYPE::UnpackedType zp = zero_point ? zero_point[bd_i].GetElem(bd_j) : 0;                         \
```

Comment:
>zero_point[bd_i].GetElem(bd_j) [](http://example.com/codeflow?start=50&length=30)

it seems that scale and zero_point share the same shape.

Reply:
The scale and zero-point inputs do have the same shape. Please refer to onnx spec: https://onnx.ai/onnx/operators/onnx__QuantizeLinear.html

Both zero-point and scale have the same shape in this code as well. The zero-point input is stored as a packed int4, so we have to get the correct 4-bit element.

Can you please clarify what you think needs to be updated?